src/main.rs

```diff
@@ -99,86 +102,160 @@ struct Args {
+    }
+}
 
-    let gbm = GbmDevice::new(gpu).expect("Failed to create GbmDevice");
+struct Window {
+    gbm_surface: Surface<BufferObject<()>>,
+    drm_display: DrmDisplay,
+}
 
-    let mut buffer_object = gbm
-        .create_buffer_object::<()>(
-            width,
-            height,
-            BufferFormat::Argb8888,
-            BufferObjectFlags::SCANOUT | BufferObjectFlags::WRITE,
-        )
-        .unwrap();
+impl Window {
+    fn new(drm_display: DrmDisplay) -> Window {
+        let format = BufferFormat::Argb8888;
+        // NOTE(mbernat): This should be used with create_surface() on drivers that support it (not nvidia)
+        // let usage = BufferObjectFlags::SCANOUT | BufferObjectFlags::RENDERING;
+        let modifiers = std::iter::once(Modifier::Linear);
+        let gbm_surface: Surface<BufferObject<()>> = drm_display
+            .gbm_device
+            .create_surface_with_modifiers(drm_display.width, drm_display.height, format, modifiers)
+            .unwrap();
+
+        Window { gbm_surface, drm_display }
+    }
 
-    let buffer_data = vec![255u8; (width * height * 4) as usize];
+    fn swap_buffers(&self) {
+        // TODO(mbernat): move this elsewhere
+        let depth_bits = 24;
+        let bits_per_pixel = 32;
+
+        // SAFETY: this must be called exactly once after `eglSwapBuffers`, which happens in `frame.finish()`.
+        let buffer_object = unsafe { self.gbm_surface.lock_front_buffer().unwrap() };
+        // TODO(mbernat): We should recycle framebuffers;
+        // one can store an FB handle in buffer object's user_data() and reuse it when it exists
+        let fb = self
+            .drm_display
+            .gbm_device
+            .add_framebuffer(&buffer_object, depth_bits, bits_per_pixel)
+            .unwrap();
+        self.drm_display
+            .gbm_device
+            .set_crtc(
+                self.drm_display.crtc.handle(),
+                Some(fb),
+                (0, 0),
+                &[self.drm_display.connector.handle()],
+                Some(self.drm_display.mode),
+            )
+            .unwrap();
+        // TODO(mbernat): Subsequent mode setting should be done with gbm_device.page_flip()
+    }
+}
 
-    buffer_object.write(&buffer_data).unwrap().unwrap();
+mod rwh_impl {
+    use super::Window;
+    use gbm::AsRaw;
+    use raw_window_handle::*;
+
+    // SAFETY: surface is valid for the duration of the program
+    unsafe impl HasRawWindowHandle for Window {
+        fn raw_window_handle(&self) -> RawWindowHandle {
+            let mut handle = GbmWindowHandle::empty();
+            handle.gbm_surface = self.gbm_surface.as_raw() as *mut _;
+            RawWindowHandle::Gbm(handle)
+        }
+    }
 
-    let depth_bits = 32;
-    let bits_per_pixel = 32;
-    let fb = gbm.add_framebuffer(&buffer_object, depth_bits, bits_per_pixel).unwrap();
+    // SAFETY: device is valid for the duration of the program
```

Comment:
What guarantees that `device` is valid for the whole program? (I believe it, just don't know where to find that info 😄)

Reply:
Good question. There is no guarantee, it just happens to be the case right now because we control everything :smile: 

If this was a library the user could easily drop the display or window and then try rendering to them.

To make this safer, one should tie together window's and handle's lifetimes. I believe `raw_window_handle` 0.6 does that but unfortunately `glutin` is on version 0.5 (as is `wgpu`; only `winit` supports the new version).

I'll add this explanation to the code.